Mathlib/Data/List/NatAntidiagonal.lean

```diff
@@ -0,0 +1,112 @@
+`list` definition we have here into `multiset` and `finset`.
+-/
+
+
+open List Function Nat
+
+namespace List
+
+namespace Nat
+
+/-- The antidiagonal of a natural number `n` is the list of pairs `(i, j)` such that `i + j = n`. -/
+def antidiagonal (n : ℕ) : List (ℕ × ℕ) :=
+  (range (n + 1)).map fun i ↦ (i, n - i)
+#align list.nat.antidiagonal List.Nat.antidiagonal
+
+/-- A pair (i, j) is contained in the antidiagonal of `n` if and only if `i + j = n`. -/
+@[simp]
+theorem mem_antidiagonal {n : ℕ} {x : ℕ × ℕ} : x ∈ antidiagonal n ↔ x.1 + x.2 = n :=
+  by
+  rw [antidiagonal, mem_map]; constructor
+  · rintro ⟨i, hi, rfl⟩
+    rw [mem_range, lt_succ_iff] at hi
+    exact add_tsub_cancel_of_le hi
+  · rintro rfl
+    refine' ⟨x.fst, _, _⟩
+    · rw [mem_range, add_assoc, lt_add_iff_pos_right]
+      exact zero_lt_succ _
+    · exact Prod.ext rfl (by simp only [add_tsub_cancel_left])
+#align list.nat.mem_antidiagonal List.Nat.mem_antidiagonal
+
+/-- The length of the antidiagonal of `n` is `n + 1`. -/
+@[simp]
+theorem length_antidiagonal (n : ℕ) : (antidiagonal n).length = n + 1 := by
+  rw [antidiagonal, length_map, length_range]
+#align list.nat.length_antidiagonal List.Nat.length_antidiagonal
+
+/-- The antidiagonal of `0` is the list `[(0, 0)]` -/
+@[simp]
+theorem antidiagonal_zero : antidiagonal 0 = [(0, 0)] :=
+  rfl
+#align list.nat.antidiagonal_zero List.Nat.antidiagonal_zero
+
+/-- The antidiagonal of `n` does not contain duplicate entries. -/
+theorem nodup_antidiagonal (n : ℕ) : Nodup (antidiagonal n) :=
+  (nodup_range _).map ((@LeftInverse.injective ℕ (ℕ × ℕ) Prod.fst fun i ↦ (i, n - i)) fun _ ↦ rfl)
+#align list.nat.nodup_antidiagonal List.Nat.nodup_antidiagonal
+
+@[simp]
+theorem antidiagonal_succ {n : ℕ} :
+    antidiagonal (n + 1) = (0, n + 1) :: (antidiagonal n).map (Prod.map Nat.succ id) :=
+  by
+  simp only [antidiagonal, range_succ_eq_map, map_cons, true_and_iff, Nat.add_succ_sub_one,
+    add_zero, id.def, eq_self_iff_true, tsub_zero, map_map, Prod.map_mk]
+  apply congr rfl (congr rfl _)
+  ext; simp
+#align list.nat.antidiagonal_succ List.Nat.antidiagonal_succ
+
+theorem antidiagonal_succ' {n : ℕ} :
+    antidiagonal (n + 1) = (antidiagonal n).map (Prod.map id Nat.succ) ++ [(n + 1, 0)] :=
+  by
+  simp only [antidiagonal, range_succ, add_tsub_cancel_left, map_append, append_assoc, tsub_self,
+    singleton_append, map_map, map]
+  congr 1
+  apply map_congr
+  simp (config := { contextual := true }) [le_of_lt, Nat.succ_eq_add_one, Nat.sub_add_comm]
+#align list.nat.antidiagonal_succ' List.Nat.antidiagonal_succ'
+
+theorem antidiagonal_succ_succ' {n : ℕ} :
+    antidiagonal (n + 2) =
+      (0, n + 2) :: (antidiagonal n).map (Prod.map Nat.succ Nat.succ) ++ [(n + 2, 0)] :=
+  by
+  rw [antidiagonal_succ']
+  simp
+  ext
+  simp
+#align list.nat.antidiagonal_succ_succ' List.Nat.antidiagonal_succ_succ'
+
+theorem map_swap_antidiagonal {n : ℕ} : (antidiagonal n).map Prod.swap = (antidiagonal n).reverse :=
+  by
```

Comment:
```suggestion
theorem map_swap_antidiagonal {n : ℕ} : (antidiagonal n).map Prod.swap = (antidiagonal n).reverse := by
```

Reply:
I left this unchanged because otherwise this line will exceeds 100 char (thus resulting in lint error).